### PR TITLE
[Merged by Bors] - feat: Prove that the measure equivalence between EuclideanSpace ℝ ι and ι → ℝ is volume preserving 

### PR DESCRIPTION
**Pull request by xroblot**
We prove that the two `MeasureSpace` structures on $\mathbb{R}^\iota$, the one coming from its identification with `ι→ ℝ` and the one coming from `EuclideanSpace ℝ ι`, agree in the sense that the measure equivalence between the two corresponding volumes is measure preserving. 

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

---

- [x] depends on: #7034 

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 1]
```diff
@@ -612,6 +612,38 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
 
 end OrthonormalBasis
 
+section EuclideanSpace
+
+variable (𝕜 ι) [IsROrC 𝕜] [Fintype ι] [DecidableEq ι]
+
+/-- The basis `Pi.basisFun` bundled as an orthormal basis of `EuclideanSpace 𝕜 ι`. -/
+noncomputable def EuclideanSpace.basisFun :
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:620]: This basis already exists elsewhere, with a different defeq; I'll try to find it when I'm next at lean!

**Review by eric-wieser — COMMENTED** [review]
I'll try to find some time to review the dependency one more time in the next few days. Thanks for getting this over the finish line!

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean (hunk 1 above)
Comment (eric-wieser) [reply Mathlib/Analysis/InnerProductSpace/PiL2.lean:620]: (it's effectively `LinearIsometryEquiv.refl`!)

Comment (xroblot) [reply Mathlib/Analysis/InnerProductSpace/PiL2.lean:620]: Thanks. I use that now to construct `EuclideanSpace.basisFun`

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 2]
```diff
@@ -612,6 +612,25 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
...
+    OrthonormalBasis ι 𝕜 (EuclideanSpace 𝕜 ι) := ⟨LinearIsometryEquiv.refl _ _⟩
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:621]: I think this might even be `default`?

Comment (eric-wieser) [reply Mathlib/Analysis/InnerProductSpace/PiL2.lean:621]: (in which case, `default` should be defined in terms of this)

Comment (xroblot) [reply Mathlib/Analysis/InnerProductSpace/PiL2.lean:617]: Indeed, this is the default OrthonormalBasis... I defined the default `OrthonormalBasis` using `EuclideanSpace.basisFun`. I hope that's what you had in mind...

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 3]
```diff
@@ -612,6 +608,29 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
...
 
+namespace EuclideanSpace
+
+variable (𝕜 ι)
+
+/-- The basis `Pi.basisFun` bundled as an orthormal basis of `EuclideanSpace 𝕜 ι`. -/
+noncomputable def basisFun :
+    OrthonormalBasis ι 𝕜 (EuclideanSpace 𝕜 ι) := ⟨LinearIsometryEquiv.refl _ _⟩
+
+theorem basisFun_repr (x : EuclideanSpace 𝕜 ι) (i : ι) :
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:619]:
```suggestion
@[simp] theorem basisFun_repr (x : EuclideanSpace 𝕜 ι) (i : ι) :
```

File: Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean [hunk 4]
```diff
@@ -293,4 +293,13 @@ protected def measurableEquiv : EuclideanSpace ℝ ι ≃ᵐ (ι → ℝ) where
...
 #align euclidean_space.coe_measurable_equiv EuclideanSpace.coe_measurableEquiv
 
+theorem coe_measurableEquiv_symm :
+    ⇑(EuclideanSpace.measurableEquiv ι).symm = (WithLp.equiv 2 _).symm := rfl
+
+theorem coe_continuousLinearEquiv :
+    ⇑(EuclideanSpace.measurableEquiv ι) = (EuclideanSpace.equiv ι ℝ) := rfl
+
+theorem coe_continuousLinearEquiv_symm :
+    ⇑(EuclideanSpace.measurableEquiv ι).symm = (EuclideanSpace.equiv ι ℝ).symm := rfl
```
Comment (eric-wieser) [inline Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean:303]: These lemmas don't actually coerce to `ContinuousLinearEquiv`, do they?

Comment (xroblot) [reply Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean:303]: I guess the name is not optimal but the idea is that `EuclideanSpace.equiv` is a the continuous linear equiv from which `EuclideanSpace.measurableEquiv` is constructed...

Comment (eric-wieser) [reply Mathlib/MeasureTheory/Measure/Haar/OfBasis.lean:303]: I'd recommend just dropping these lemmas; they follow by transitivity of the two lemmas above, and the "simplifying" lemma that says `EuclideanSpace.equiv` is just `WithLp.equiv`

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 5]
```diff
@@ -612,6 +608,31 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
 
 end OrthonormalBasis
 
+namespace EuclideanSpace
+
+variable (𝕜 ι)
+
+/-- The basis `Pi.basisFun` bundled as an orthormal basis of `EuclideanSpace 𝕜 ι`. -/
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:615]:
```suggestion
/-- The basis `Pi.basisFun`, bundled as an orthormal basis of `EuclideanSpace 𝕜 ι`. -/
```
Can you add a lemma that relates this definition with [`PiLp.basisFun`](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/NormedSpace/PiLp.html#PiLp.basisFun)?

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 6]
```diff
@@ -612,6 +608,31 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
...
+noncomputable def basisFun :
+    OrthonormalBasis ι 𝕜 (EuclideanSpace 𝕜 ι) := ⟨LinearIsometryEquiv.refl _ _⟩
+
+@[simp]
+theorem basisFun_repr (x : EuclideanSpace 𝕜 ι) (i : ι) :
+    (basisFun ι 𝕜).repr x i = x i := rfl
+
+theorem basisFun_toBasis :
+    (basisFun ι 𝕜).toBasis =
+      (Pi.basisFun 𝕜 ι).map (EuclideanSpace.equiv ι 𝕜).toLinearEquiv.symm := rfl
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:625]:
```suggestion
    (basisFun ι 𝕜).toBasis = PiLp.basisFun _ 𝕜 ι := rfl
```

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 7]
```diff
@@ -612,6 +608,30 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
 
 end OrthonormalBasis
 
+namespace EuclideanSpace
+
+variable (𝕜 ι)
+
+/-- The basis `Pi.basisFun`, bundled as an orthormal basis of `EuclideanSpace 𝕜 ι`. -/
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:615]:
```suggestion
/-- The basis `Pi.basisFun`, bundled as an orthonormal basis of `EuclideanSpace 𝕜 ι`. -/
```

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 8]
```diff
@@ -612,6 +608,30 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
...
+noncomputable def basisFun :
+    OrthonormalBasis ι 𝕜 (EuclideanSpace 𝕜 ι) := ⟨LinearIsometryEquiv.refl _ _⟩
+
+@[simp]
+theorem basisFun_repr (x : EuclideanSpace 𝕜 ι) (i : ι) :
+    (basisFun ι 𝕜).repr x i = x i := rfl
+
+theorem basisFun_toBasis : (basisFun ι 𝕜).toBasis = PiLp.basisFun _ 𝕜 ι := rfl
+--      (Pi.basisFun 𝕜 ι).map (EuclideanSpace.equiv ι 𝕜).toLinearEquiv.symm := rfl
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:624]: This can be deleted

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 9]
```diff
@@ -612,6 +608,30 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
...
+
+variable (𝕜 ι)
+
+/-- The basis `Pi.basisFun`, bundled as an orthormal basis of `EuclideanSpace 𝕜 ι`. -/
+noncomputable def basisFun :
+    OrthonormalBasis ι 𝕜 (EuclideanSpace 𝕜 ι) := ⟨LinearIsometryEquiv.refl _ _⟩
+
+@[simp]
+theorem basisFun_repr (x : EuclideanSpace 𝕜 ι) (i : ι) :
+    (basisFun ι 𝕜).repr x i = x i := rfl
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:621]: Can you add the lemma about basisFun ι 𝕜 i` too

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean (hunk 8 above)
Comment (xroblot) [reply Mathlib/Analysis/InnerProductSpace/PiL2.lean:624]: Oops, sorry about that...

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean (hunk 9 above)
Comment (xroblot) [reply Mathlib/Analysis/InnerProductSpace/PiL2.lean:621]: I added the version that's defeq to `Pi.basisFun ι 𝕜 i`. Another possibility would be to defeq to `PiLp.basisFun` but I think it is better to go directly to the simplest expression.

File: Mathlib/Analysis/InnerProductSpace/PiL2.lean [hunk 10]
```diff
@@ -612,6 +608,33 @@ protected theorem repr_reindex (b : OrthonormalBasis ι 𝕜 E) (e : ι ≃ ι')
...
+
+variable (𝕜 ι)
+
+/-- The basis `Pi.basisFun`, bundled as an orthornormal basis of `EuclideanSpace 𝕜 ι`. -/
+noncomputable def basisFun :
+    OrthonormalBasis ι 𝕜 (EuclideanSpace 𝕜 ι) := ⟨LinearIsometryEquiv.refl _ _⟩
+
+@[simp]
+theorem basisFun_apply (i : ι) :
+      basisFun ι 𝕜 i = Pi.basisFun 𝕜 ι i := rfl
```
Comment (eric-wieser) [inline Mathlib/Analysis/InnerProductSpace/PiL2.lean:621]: This one is badly typed, I'll push a fix.

**Review by eric-wieser — APPROVED** [review]
bors d+

Thanks! I added a few shortcut lemmas; let me know if you agree with them.

**Comment by xroblot** [comment]
LGTM. Thanks!

bors r+

**Pull request closed without merging**